Outline10

speudusa opened the pull request:
The Methods Chapter.  bum ba ba! *jazz hands*

This was mostly from scratch, so any suggestions about overall flow, language, or complexity of code blocks are welcome.  I got myself stuck overthinking the studio.  Is this too simple?   I was hoping that this would fill in the basics of methods, so that when we are able to use them to instantiate objects we can focus more on access modifiers and calling between classes instead of worrying about signature and invocation 101 stuff.

CFreyLC committed:
created methods chapter.  index and intro

CFreyLC committed:
added sections 2, 3, and 4; &&  replit examples with links

CFreyLC committed:
worked on recursion, anon methods, and scope

CFreyLC committed:
wrapped up scope, recursion.  revised chapter 1x.  updated formatting in strings chapter

CFreyLC committed:
added exercises and optional arguments/parameters

CFreyLC committed:
added studio, reviewed and revised all other sections: grammar, readability

CFreyLC committed:
updated index no more conflicts

speudusa commented:
I hope i didn't add too much extra.  you've already looked at the loops...  i was in an merge flow after assignment 1.

CFreyLC committed:
code block updates

CarlyLanglois commented on src/chapters/methods/method-signatures.rst:
```diff
@@ -0,0 +1,156 @@
+In this example, we are using ``static`` as our **access modifier**.  This is lets C# know that this method
+belongs to the current class it is contained in and not to expect anything **new** to be created.  
+
+Our **return type** is **int**.  This lets you know how to use the method.  If you were expecting a **double** then you will needing
```
If you expect a **double**, then you will need* ...

CarlyLanglois commented on src/chapters/methods/method-signatures.rst:
```diff
@@ -0,0 +1,156 @@
+
+When writing your own methods, provide as many clues as possible to help with readability.  
+The **method body** shows you how the input will be manipulated within the method.  
+Here it will be used to square itself.  This method will ``return`` a value which we can use else where in our code.  Since this method is static, 
```
elsewhere is one word

CarlyLanglois commented on src/chapters/methods/method-signatures.rst:
```diff
@@ -0,0 +1,156 @@
+
+You should be fairly familiar with the replit IDE by now.  
+
+- **Line 1** us where the ``using`` statement goes.  
```
tells us?

CarlyLanglois commented on src/chapters/methods/method-signatures.rst:
```diff
@@ -0,0 +1,156 @@
+- **Line3** introduces the ``MainClass`` class followed by ``{``.  The closing ``}`` is in **Line 21**.  
+- Starting at **Line 11** is the Main method statement, followed by a ``{`` which closes on **Line 20**.
+
+We can place our method anywhere in the ``MainClass`` but outside the Main method.  In this example, we placed our
```
Im not crazy about the way this first statement is phrased. Can you rephrase it to be more explicit that this method can be anywhere in the class as long as it is not nested within the Main method's method body?

CarlyLanglois commented on src/chapters/methods/method-signatures.rst:
```diff
@@ -0,0 +1,156 @@
+
+We are using static methods with expected returns, these take input and return output, or values.  So we must provide them with input if we want to see them work.  
+Since this method is returning a numeric value, it needs to be stored somewhere.  
+You can create a variable to do so, as we did in **Line 14** with the variable ``squaredNum``.  
```
i think this is actually in line 13 now. And actually, this info could be more digestible by referencing the code snippet rather than the line #s - since the reader needs to scroll up or flip to the replit page to confirm what code exists on each line.

Check the line #s in the rest of this section as well.

Also, the repl linked here contains different code than what is discussed

CarlyLanglois commented on src/chapters/methods/method-signatures.rst:
```diff
@@ -0,0 +1,156 @@
+   #. The method name should be ``perimeter``.
+   #. A return statement is inappropriate for this type of method.
+
+   .. ans: a.
```
Im not getting why a is the correct answer here :/

CarlyLanglois commented on src/chapters/methods/using-methods.rst:
```diff
@@ -0,0 +1,294 @@
+
+.. admonition:: Example
+
+   .. sourcecode:: csharp
```
this example would do well with a repl linked as well so that the students have another opportunity to see how these two methods need to be defined within a class

CarlyLanglois commented on src/chapters/methods/using-methods.rst:
```diff
@@ -0,0 +1,294 @@
+
+In this example, we are asking the method to print a friend's name and birth date inside a preprogrammed message.  
+This might see really simple, but if you had a list of names to run through, calling your method for each is a LOT less typing than 
+a ``Console.WriteLine`` statement with you message for each one.  That would get old fast.  
```
your* message

CarlyLanglois commented on src/chapters/methods/using-methods.rst:
```diff
@@ -0,0 +1,294 @@
+For our human understanding we gave the parameters helpful names, but those names are NOT variables or values.  
+They are holding space for your actual arguments.  The method will take your arguments and use them as designated by the parameters.
+
+When we call the method, we can pass our arguements directly or via variables and check the output to see if we are happy with ther results.
```
the* results

CarlyLanglois replied:
arguments*

CarlyLanglois commented on src/chapters/methods/using-methods.rst:
```diff
@@ -0,0 +1,294 @@
+
+In the example we called BirthdayNotes twice.  The first time, ``BirthdayNotes("Willow", 8, 3)`` and the second time ``BirthdayNotes("This is a string, too", 3-5, 987654321)``.  
+The code ran both times because the data types of the arguments matched those of the parameters.  The second call doesn't make any sense in the context
+of the method's overall function.  However, since it was provided with arguments of the same data type as the parameters, the method ran as expected.  
```
maybe instead: "The second call doesn't make any sense in the context
of the method's overall job."

only to avoid using "function" as a description of what the method does

CarlyLanglois commented on src/chapters/methods/using-methods.rst:
```diff
@@ -0,0 +1,294 @@
+         Console.WriteLine(total1);
+         
+         }
+      }
```
can you un-indent this block so its on the same spacing as the method above?

CarlyLanglois commented on src/chapters/methods/using-methods.rst:
```diff
@@ -0,0 +1,294 @@
+Optional Arguments
+^^^^^^^^^^^^^^^^^^^^
+
+An argument becomes optional when the parameters is coded to take a **default value**. 
```
singular parameters

CarlyLanglois commented on src/chapters/methods/method-recursion.rst:
```diff
@@ -0,0 +1,111 @@
+
+If methods can use methods, can a method use itself?  
+
+The answer is yes.  This process is called **recursion** and is 
```
make "this process..." into 2 sentences for clarity on the definition of recursion

CarlyLanglois commented on src/chapters/methods/method-recursion.rst:
```diff
@@ -0,0 +1,111 @@
+
+         class MainClass {
+         
+            static double FactorialRecursive(double number)
```
check your indentations here too

CarlyLanglois commented on src/chapters/methods/method-recursion.rst:
```diff
@@ -0,0 +1,111 @@
+
+Here is an example of a recursive method.  In **Lines 5-12** we create the function.  
+The base case is the ``if`` statement in **Line 7** which says that when the argument reaches ``0``, the method will return ``1`` and stop.
+But if the arguement is not ``0``, we leave the ``if`` statment and go to **Line 11** which uses the method to call itself once more, with modified parameters.
```
make this 2 sentences

CarlyLanglois commented on src/chapters/methods/method-recursion.rst:
```diff
@@ -0,0 +1,111 @@
+
+         class MainClass {
+         
+            static double FactorialLoop(double number)
```
indentation is off on this example too. maybe you should check them all :)

CarlyLanglois replied:
this is a great example btw

CarlyLanglois commented on src/chapters/methods/method-scope.rst:
```diff
@@ -0,0 +1,186 @@
+**Scope** refers to visibility of a particular variable within your code.  
+So far we have defined variables directly in the Main method within the ``MainClass`` in the replit IDE, 
+but also inside loops and methods.  Perhaps you have exerienced an error when trying to print or call a variable that 
+was created inside you loop from outside of the loop.  This is an example of a scope issue.  
```
your* loop

CarlyLanglois commented on src/chapters/methods/method-scope.rst:
```diff
@@ -0,0 +1,186 @@
+When you see this message, you might need to backstep through your work and look at your boxes.  
+*Did you provide a way to open the lid?  Did you mean to work within the loop or method?  
+Did you really want* that *access modifier?  Did you accidently open the lid when it should stay closed?*
+Scope can vary from too narrow, as in trying to access a variable created within a loop or to broad, 
```
too* broad

CarlyLanglois replied:
also this sentence is too long. can you make it into 2 ... maybe 3

CarlyLanglois commented on src/chapters/methods/method-scope.rst:
```diff
@@ -0,0 +1,186 @@
+
+      class MainClass {
+
+         static int SumItUp(int num1, int num2, int num3)
+         {
+            return num1+ num2+ num3;
+         }
+
+         public static void Main (string[] args) {
+
+            int total = SumItUp(1,2,3);
+            Console.WriteLine(total);
+         }
+      }
```
ok this note is kind of nit picky but try keeping the bracket usage consistent. here we have the class and the main method with an opening bracket on the same line as the definition yet the SumItUp method's opening bracket is on the next line. 
I don't care which you use - although I think the unit 2 book uses almost exclusively opening brackets on the following line

speudusa replied:
i was matching replit, but VS does put the bracket below like the examples in unit 2.

CarlyLanglois commented on src/chapters/methods/method-scope.rst:
```diff
@@ -0,0 +1,186 @@
+----------------------
+
+Class level scope you are already familiar with, but once we begin to add *more* classes to our code, this can become an issue. 
+When we get to classes, we will discuss how to handle class level scope in more detail. (Hint: think access modifiers).
```
are they familiar with it yet? this may change depending where this chapter gets placed in lesson order

CarlyLanglois commented on src/chapters/methods/method-scope.rst:
```diff
@@ -0,0 +1,186 @@
+Method Level Scope 
+------------------------
+
+With method level scope, any variables declared **within** a method are not avaliable outside the method.
```
within shouldn't be boldened here because it is not a vocabulary word. you can use just one set of asterisks like *this* to make the text italicized for emphasis however

CarlyLanglois commented on src/chapters/methods/method-scope.rst:
```diff
@@ -0,0 +1,186 @@
+Moving down to **Line 14**, we see another ``for`` loop.  This loop will print each value of ``j`` from within its codeblock as is. 
+There are no scope issues here.  Run the code and you will see ``10, 13, 16, 19`` print to the console on new lines.
+
+**Line 20** is where we find a scope issue.  We want to know what ``j`` is **outside** the ``for`` loop.
```
I would italics here again for the emphasis on *outside*

CarlyLanglois commented on src/chapters/methods/method-scope.rst:
```diff
@@ -0,0 +1,186 @@
+There are no scope issues here.  Run the code and you will see ``10, 13, 16, 19`` print to the console on new lines.
+
+**Line 20** is where we find a scope issue.  We want to know what ``j`` is **outside** the ``for`` loop.
+However, ``j`` doesn't exist outside of the context of the ``for`` loop, so the ``Console.WriteLine`` statement in **Line 20**
```
try adding: "However, ``j`` doesn't exist outside of the context of the ``for`` loop block ..."

to help make it more clear what a code block is

CarlyLanglois commented on src/chapters/methods/method-scope.rst:
```diff
@@ -0,0 +1,186 @@
+
+.. admonition:: Question
+
+   Identify the level of scope for each variable.  Bonus points if you can explain what will happen **Lines 16 & 17** and why.
```
kind of annoying but I'd take out "bonus point for" and just ask them to consider this on their own. since we're not giving point for concept check questions, we would almost certainly get questions about this from students

CarlyLanglois commented on src/chapters/methods/method-scope.rst:
```diff
@@ -0,0 +1,186 @@
+   #. Method scope
+   #. Loop scope
+
+.. ans: 1 block, 2 method, 3 class
```
a few issues with this question - a) the answer options dont match the answer :) and b) the container class isnt visible so a class level scope can be determined by deduction but really showing the container class would be a good fit here.

I'd also consider just asking: 'what is the scope level of the the variable `int byHalf`

CarlyLanglois commented on src/chapters/methods/exercises.rst:
```diff
@@ -0,0 +1,188 @@
+.. admonition:: Note
+
+   As you build your methods, remember to keep the output in mind.  
+   Do you want a value returned or are do you want the method to print directly to the console when invoked?
```
or are do you want - ?!

CarlyLanglois commented on src/chapters/methods/exercises.rst:
```diff
@@ -0,0 +1,188 @@
+   Do you want a value returned or are do you want the method to print directly to the console when invoked?
+   How many things are being repeated?
+
+There is no starter code for these exercises, so create a new C# project
```
normally, i'd support this approach but I think we should give them a starter repl with the container class and main method created since we want to focus of the practice to be on the method creation. what do you think?

CarlyLanglois commented on src/chapters/methods/exercises.rst:
```diff
@@ -0,0 +1,188 @@
+
+Asthetically it's not a "perfect" square, but a single argument led to its creation.
+
+#. Write a method ``MakeRectangle(width, height)`` that returns a
```
this can be modified to be called 3. (right now it renders as 1. because of the text line breaking this up into 2 sections)

CarlyLanglois commented on src/chapters/methods/exercises.rst:
```diff
@@ -0,0 +1,188 @@
+         ###
+          #
+
+   .. tip::
+
+      Consider what happens if you create a triangle and reverse it within the same method?
```
im not getting the idea behind this tip. can you elaborate? or maybe just get rid of it?

speudusa replied:
This was from the JS book.  I didn't really use it when working on solutions.  I left it in case someone found it helpful.  I;m not very attached to it, so I have no issue pulling it.

CarlyLanglois commented on src/chapters/methods/exercises.rst:
```diff
@@ -0,0 +1,188 @@
+#. Write a method ``MakeLine(size)`` that returns a line with exactly ``size``
+   hashes.
+
+   .. sourcecode:: csharp
+
+      MakeLine(5);
+
+   **Console Output**
+
+   ::
+
+      #####
```
we should provide the signature for at least the first of these methods to get students into the groove. also though, if we're wanting these methods to return these shapes, then the "Console Output" will not be anything unless Console.WriteLine(MakeLine(5)) is called, right?

Maybe we should ask the students instead to write methods that return void but print the shapes instead? having the shapes be returned would require more practice with return types though

speudusa replied:
does this mean we should give them the first method?  or the method template?  

i used void methods.

CarlyLanglois commented on src/chapters/methods/studio.rst:
```diff
@@ -0,0 +1,88 @@
+Warm-Up: Reverse a Number Array
+--------------------------------
+
+1. Define a method called ``ReverseNumbers`` that will sort the numbers within an array in descending order.
```
here's where you can say: create a new repl ... yada yada ... don't forget to create a container ``MainClass`` and copy the ``Main`` method signature so that you can run the program and call your ``ReverseNumbers`` method

CarlyLanglois commented on src/chapters/methods/studio.rst:
```diff
@@ -0,0 +1,88 @@
+Studio: Methods
+====================
+
+The ``Reverse`` method flips the order of the elements within an array.
```
actually, having a note in the intro to the chapter on these built in array methods and how they have seen some methods before would be a great way to ease the reader into the new topic of what methods are. you can link back to the array methods appendix

CarlyLanglois commented on src/chapters/methods/studio.rst:
```diff
@@ -0,0 +1,88 @@
+2. Within the method, ``Sort`` the array.
+3. Use ``Reverse`` method to put the items into descending order.  
+4. Using a loop of your choice, print your final array to the console.
+5. Be sure to test your method, call it at least twice with arrays of differnt sizes and numbers.  
```
different*

CarlyLanglois commented on src/chapters/methods/studio.rst:
```diff
@@ -0,0 +1,88 @@
+
+1. Define a method called ``ReverseLetters`` that will reverse each character inside a string and print the results to the console.  This will be similar to the ``ReverseNumbers`` method.
+2.  Look through the `String Methods <string-methods>`_  and `Array Methods <array-methods>`_ to figure out the best approach for this.  Consider the following: How can you access or alter characters within strings? What data types are going into your method, being used in your method, and printed by your method?
+3. Be sure to test your method by invoking it a few times.  Some test examples you could use: ``C Sharp``, ``Hello``, ``Coding is fun!`` they should return as ``prahS C``, ``olleH``, and ``!nuf so gniodC`` .
```
the last reverse would be : ``!nuf si gnidoC``

CarlyLanglois commented on src/chapters/methods/studio.rst:
```diff
@@ -0,0 +1,88 @@
+--------------------
+
+Now that you can completely reverse a string, let's use this to check to see if a word is a palindrome.
+A palindrome is a words that is spelled the same forwards and backwards.  Some examples are "tacocat", "kayak", and racecar".
```
and "racecar" - it's missing the opening quotation mark

CarlyLanglois commented on src/chapters/methods/studio.rst:
```diff
@@ -0,0 +1,88 @@
+
+
+You have alredy built a function that reverses a string.  You could extend that method to be able to detect palindromes. 
+Or since, this data type for this will always be a string, you could build it using only string methods and a loop. 
```
"Or since, this data type for this will always be a string," this construction is throwing me off.
Can you say instead: "Or you could build it using only string methods and a loop." Do we lose any critical info like this?

CarlyLanglois commented on src/chapters/methods/studio.rst:
```diff
@@ -0,0 +1,88 @@
+Or since, this data type for this will always be a string, you could build it using only string methods and a loop. 
+The choice is up to you.  
+
+1.  Define your method ``IsPalindrome``, to check if a string is equal to its reverse.  Think what data type the overall **outcome** will require.
```
i'd keep outcome as plain text (not bold)

CarlyLanglois reviewed (approved):
Overall really great content here Courtney! The notes we talked about this morning I think can mainly be addressed in the introductory page. Let's add some language around these ideas:

a) methods vs functions: id say, lets tell the students that technically c# programmers use only methods because methods MUST be defined within a class. but at times in the book and in conversations, its ok to interchange methods and functions
b) calling/invoking - invoking is defined on one of the first pages but you use the phrase 'call a method' a few times before its introduced or defined at all. can you add a quick definition of what it means to call a method. something to distinguish between the act of writing and defining the method and then using it/running it
c) the static/instance method definitions are a little off. from what we talked about this morning - can you do some slight refactoring of these pages to instead simply let the students know that methods have 'signatures' and we'll expand upon what a method signature requires in the classes chapters?

otherwise - the studio is an appropriate amount of work yes :)

CFreyLC committed:
revisions on sections 1-3. started section 4.

CFreyLC committed:
sections 4 and 5, exercises, and studio revised.

CFreyLC committed:
language updates in exercises and studio

CFreyLC committed:
created classes/objects section

CFreyLC committed:
language update in class/object section

CFreyLC committed:
pulled latest main into feature branch

CFreyLC committed:
created new chapter for quick class intro, and math objects

CFreyLC committed:
created math methods appendix.  revisions for class-object chapter

speudusa commented:
made corrections based on the revisions provided.  Changed formatting of example presentation, language.  Created new chapter for basic class and object intro.  Tacked the Math Class on the end.  Didn't do any exercises or studio with this.  Classes will be easier in VS, and we'll get a lot of practice in the Class chapters.

I updated some of the Check Your Understanding questions.  Once this gets the go ahead to merge, i'll update Canvas with the new questions.

CarlyLanglois commented on src/chapters/methods/what-is-a-method.rst:
```diff
@@ -0,0 +1,123 @@
+
+.. admonition:: Note
+
+   In C# or other OOP languages, developers use methods because they are assiociated with a class.
```
OOP is not yet defined - otherwise great note!

CarlyLanglois commented on src/chapters/methods/method-signatures.rst:
```diff
@@ -0,0 +1,211 @@
+The reason for this is that we are working within a single class.  
+If we had multiple classes this would be more significant.  We will discuss this more when we get to class and class members.
+
+We are using the ``static`` modifier.   This is lets C# know that this method
```
This lets C#* ...

CarlyLanglois commented on src/chapters/methods/method-signatures.rst:
```diff
@@ -0,0 +1,211 @@
+   .. sourcecode:: csharp
+      :linenos:
+         
+         static int Doubles(int x)
```
i think the indentation on this codeblock is off

CarlyLanglois commented on src/chapters/methods/using-methods.rst:
```diff
@@ -0,0 +1,302 @@
+      
+
+In this example, we are asking the method to print a friend's name and birth date inside a preprogrammed message.  
+This might see really simple, but if you had a list of names to run through, calling your method for each is a LOT less typing than 
```
This might seem*

CarlyLanglois commented on src/chapters/methods/method-scope.rst:
```diff
@@ -0,0 +1,194 @@
+      class MainClass 
+      {
+
+      static void PrintThis()
```
check the indentation levels here too

CarlyLanglois commented on src/chapters/classes-objects-math/class-type.rst:
```diff
@@ -0,0 +1,116 @@
+Class Type
+-------------------
+
+Quick refesher on data types.  As discussed in chapter 4, C# contains value types and reference types.  
```
can you replace the reference to chapter 4 with a reference to the section title - and you can even use a :ref: link to give a hyperlink back to that page.
we want to avoid referring to chapter numbers because if we move sections around, we dont want to have to check the book for references like this

CarlyLanglois commented on src/chapters/classes-objects-math/classes-objects-methods.rst:
```diff
@@ -0,0 +1,363 @@
+   When working with replit.com, you will need to create a new class in a new file.  
+   In order for your class to be recognized by ``MainClass`` in replit, you will need to provide your class its own unique ``namespace``.
+   
+   We will be moving to a differnt IDE soon, so any replit.com code examples will be provided to you.  
```
different*

CarlyLanglois commented on src/chapters/classes-objects-math/classes-objects-methods.rst:
```diff
@@ -0,0 +1,363 @@
+   We will be moving to a differnt IDE soon, so any replit.com code examples will be provided to you.  
+
+When working in a project with multiple classes, it is common to have a ``namespace`` that connects all classes within a project.
+In this context, a ``namespace`` can be more for organizational purposes, as in it will provide direction to each class within 
```
Can you convert this to two sentences?

CarlyLanglois commented on src/chapters/classes-objects-math/classes-objects-methods.rst:
```diff
@@ -0,0 +1,363 @@
+When working in a project with multiple classes, it is common to have a ``namespace`` that connects all classes within a project.
+In this context, a ``namespace`` can be more for organizational purposes, as in it will provide direction to each class within 
+the main project.  If the ``namespace`` is not part of the class that holds your ``Main method`` you will have to access it through a ``using`` statement.
+Just like we did will Collections.  This tells the compiler to look through any classes within the same ``namespace`` for definitions and requirements.
```
with* Collections

CarlyLanglois commented on src/chapters/classes-objects-math/classes-objects-methods.rst:
```diff
@@ -0,0 +1,363 @@
+To delcare a *new* object of a class you will use the keyword ``new`` as you have done with ``Lists``, arrays, and dictionaries.  
+The data type for these objects is the class, which you reference directly.
+
+.. sourcecode:: csharp
+
+ //data type                  constructor with possilbe parameters
+   ClassName objectName = new ClassName(possible parameters); 
+
+
+In our ``Cat`` class example, we instantiate ``myCat`` which is a new ``Cat`` object.
+
+.. sourcecode:: csharp
+   :lineno-start: 10
+
+      Cat myCat = new Cat();         
```
more indentations to check here

CarlyLanglois commented on src/chapters/classes-objects-math/classes-objects-methods.rst:
```diff
@@ -0,0 +1,363 @@
+^^^^^^^^^^^^^^^^^^^^^^^
+
+When we instantiate an object, we use the class **constructor** to provide any needed values to initialize any attributes.  
+The syntax just like the parameter list of a method.
```
this is not a complete sentence oups! not sure what you're conveying here

CarlyLanglois commented on src/chapters/classes-objects-math/classes-objects-methods.rst:
```diff
@@ -0,0 +1,363 @@
+Class definitions deterimine any **attributes** the object requires, as well as any **behaviors** of the object.
+Attributes are class variables, also called **fields**.  They can be hard coded, or filled in at the time of instantiation.
+
+Methods are the behvaiors or actions your class is able to preform on objects created by that class.  
```
behaviors* and perform*

CarlyLanglois commented on src/chapters/classes-objects-math/classes-objects-methods.rst:
```diff
@@ -0,0 +1,363 @@
+
+
+Since we used the default constructor to create the ``myCat`` object, 
+the attribute ``name`` will "Alyce" or exactly what we coded it to be.  
```
I think you're missing a verb in this sentence - will be*?

CarlyLanglois commented on src/chapters/classes-objects-math/math-class.rst:
```diff
@@ -0,0 +1,159 @@
+Having a built in ``Math`` class can be very helpful when it comes to common math elements, such as ``PI``.  
+
+One of the benefits of constant values is that you don't have to memorize them, or define them in a variable.
+C# stores the value of pi upto 16 decimal places for us.  
```
up to - needs to be two words

CarlyLanglois reviewed (approved):
excellent additions! The methods revisions are really well done - thanks for going back and making those changes. 
I'm digging this classes and objects introduction as well - I suspect we may have some overlap with the classes pt 1 section but we'll return to that talk when you move those pages over i suppose.

CFreyLC committed:
language revisions
